scopesim/server/gitdir.py

```diff
@@ -0,0 +1,133 @@
+        else:
+            dir_out = os.path.join(output_dir, "/".join(download_dirs.split("/")[:-1]))
+    else:
+        dir_out = output_dir
+
+    try:
+        opener = urllib.request.build_opener()
+        opener.addheaders = [('User-agent', 'Mozilla/5.0')]
+        urllib.request.install_opener(opener)
+        response = urllib.request.urlretrieve(api_url)
+    except KeyboardInterrupt:
+        # when CTRL+C is pressed during the execution of this script,
+        logging.error("GitHub download interrupted by User")
+        sys.exit()
+
+    if not flatten:
+        # make a directory with the name which is taken from the actual repo
+        os.makedirs(dir_out, exist_ok=True)
+
+    # total files count
+    total_files = 0
+
+    with open(response[0], "r") as f:
+        data = json.load(f)
+        # getting the total number of files so that we
+        # can use it for the output information later
+        total_files += len(data)
+
+        # If the data is a file, download it as one.
+        if isinstance(data, dict) and data["type"] == "file":
+            try:
+                # download the file
+                opener = urllib.request.build_opener()
+                opener.addheaders = [('User-agent', 'Mozilla/5.0')]
+                urllib.request.install_opener(opener)
+                urllib.request.urlretrieve(data["download_url"], os.path.join(dir_out, data["name"]))
+                logging.info(f"Downloaded: {data['name']}")
+
+                return total_files
+            except KeyboardInterrupt:
+                # when CTRL+C is pressed during the execution of this script,
+                logging.error("GitHub download interrupted by User")
+                sys.exit()
+
+        for file in data:
+            file_url = file["download_url"]
+            file_name = file["name"]
+            file_path = os.path.join(dir_out, file["path"])
+
+            if flatten:
+                path = os.path.basename(file_path)
+            else:
+                path = file_path
+            dirname = os.path.dirname(path)
+
+            if dirname != '':
+                os.makedirs(os.path.dirname(path), exist_ok=True)
+                # os.makedirs(os.path.join(dir_out, os.path.dirname(path)), exist_ok=True)
+            else:
+                pass
+
+            if file_url is not None:
+                try:
+                    opener = urllib.request.build_opener()
+                    opener.addheaders = [('User-agent', 'Mozilla/5.0')]
+                    urllib.request.install_opener(opener)
+                    # download the file
+                    urllib.request.urlretrieve(file_url, path)
+                    logging.info(f"Downloaded: {file['path']}")
+
+                except KeyboardInterrupt:
+                    # when CTRL+C is pressed during the execution of this script,
+                    logging.error("GitHub download interrupted by User")
+                    sys.exit()
```

Comment:
It seems un-Pythonic to just exit! Suggestion to send the Exception upstream (which gives the user the chance to catch it themselves)
```suggestion
                    raise
```

Reply:
Similar for all the other `sys.exit()`s